src/python/pants/invalidation/build_invalidator.py

```diff
@@ -77,10 +77,13 @@ def key_for_target(self, target, transitive=False, fingerprint_strategy=None):
     hasher = hashlib.sha1()
     hasher.update(self._cache_key_gen_version)
     key_suffix = hasher.hexdigest()[:12]
-    if transitive:
-      target_key = target.transitive_invalidation_hash(fingerprint_strategy)
+    if direct_checker and direct_checker(target):
```

Comment:
hmm, maybe not an abstract one, the method in base class can return False. Otherwise, all derived classes has to implement this, but only 1 actually overrides it.